File: Week_01/189-rotate-array.java
class Solution2 {
    public void rotate(int[] nums, int k) {
        int count = 0;
        for(int start = 0; count < nums.length; start++) {
            int current = start;
            int prev = nums[start];
            do {
                int next = (current + k) % nums.length;
                int temp = nums[next];
                nums[next] = prev;
                prev = temp;
                current = next;
                count++;
            }while(start != current);
        }
    }

    public static void main(String[] args) {
        int[] nums = {-1, -100, 3, 99};
        Solution2 solution = new Solution2();
        solution.rotate(nums, 2);
    }
}

File: Week_01/189-rotate-array.cpp
#include <vector>

using namespace std;

class Solution {
public: 
    void rotate(vector<int>& nums, int k) {
        reverse(nums, 0, nums.size());
        reverse(nums, 0, k);
        reverse(nums, k, nums.size());
    }

    int main(int argc, char const *argv[])
    {
        vector<int> nums(7);
        for (size_t i = 0; i < 8; i++)
        {
            nums[i] = i;
        }

        rotate(nums, 3);
        return 0;
    }
    

private:
    void reverse(vector<int>& nums, int start, int end) {
        for (size_t i = start; i < (end - start) / 2; i++)
        {
            int temp = nums[start + i];
            nums[start + i] = nums[end - i - 1];
            nums[end - i - 1] = temp;
        }
        
    }
};